cpp/src/arrow/compute/function_internal.h

```diff
@@ -352,6 +352,14 @@ static inline Result<std::shared_ptr<Scalar>> GenericToScalar(
   return MakeNullScalar(value);
 }
 
+static inline Result<std::shared_ptr<Scalar>> GenericToScalar(
```

Comment:
Is this specialization necessary? I think we already have `std::optional<>` and `std::shared_ptr<DataType>` specialized, would them be enough for `std::optional<std::shared_ptr<DataType>`?